### PR TITLE
Issue/2060 docs compiler section

### LINKED ISSUE #2060: add compiler section to docs
Subsections:
- compiler flags: not sure if this is required since the `--help` flag can just be used.
- API:
  - external:
    - compiler json export for automated tooling (#2048)
  - internal:
    - `inmanta.compiler.data` once it contains more data
    - `inmanta.ast.type`
    - `inmanta.execute.proxy.DynamicProxy` `return_value` and `unwrap` (#2047)
    - ...

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Adds some documentation on the API offered by the compiler, internal as well as the exported JSON (`inmanta compile --json`).

closes #2060

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] ~~Type annotations are present~~
- [x] ~~Code is clear and sufficiently documented~~
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] ~~Sufficient test cases (reproduces the bug/tests the requested feature)~~
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: docs/reference/json.rst [hunk 1]
```diff
@@ -0,0 +1,44 @@
+Inmanta JSON Reference
+======================
+
+This page documents the compile data output when compiling with the `--json` flag. The structure
+of this JSON is defined by :py:class:`inmanta.compiler.data.ExportCompileData` which inherits
+from :py:class:`pydantic.BaseModel`. To produce the JSON, `model.json()` is called. See the
```
Comment (arnaudsjs) [inline docs/reference/json.rst:6]:
```suggestion
from :py:class:`pydantic.BaseModel`. To produce the JSON representation of the object, `model.json()` is called. See the
```

File: src/inmanta/ast/export.py [hunk 2]
```diff
@@ -52,24 +54,59 @@ class Range(BaseModel):
...
 
 
 class ErrorCategory(str, Enum):
+    """
+        Category of an error.
+    """
+
     plugin = "plugin_exception"
+    """
+        Explicitly raised :py:class`inmanta.plugins.PluginException` by a plugin.
```
Comment (arnaudsjs) [inline src/inmanta/ast/export.py:72]:
```suggestion
        A plugin explicitly raised an :py:class:`inmanta.plugins.PluginException`.
```

File: src/inmanta/ast/type.py [hunk 3]
```diff
@@ -79,23 +80,20 @@ def get_resolver_for(self, namespace: Namespace):
...
-        Inmanta instances are represented by instances of other Python types.
+        This class is the abstract base class for all types in the Inmanta :term:`DSL` that represent basic data. These are
+        types that are not relations. Instances of subclasses represent a type in the Inmanta language.
     """
 
     def validate(self, value: Optional[object]) -> bool:
         """
-            Validate the given value to check if it satisfies the constraints
-            associated with this type
+            Validate the given value to check if it satisfies the constraints associated with this type. Returns true iff
```
Comment (arnaudsjs) [inline src/inmanta/ast/type.py:89]:
```suggestion
            Validate the given value to check if it satisfies the constraints associated with this type. Returns true if
```

File: docs/reference/api.rst [hunk 4]
```diff
@@ -1,6 +1,31 @@
 Inmanta API reference
 =====================
 
+This page describes parts of the compiler that provide a stable API that could be used from modules.
+
+.. warning::
+    Only those parts explicitly mentioned here are part of the internal API. They provide a stable
+    interface. Other parts of the containing modules provide no such guarantees.
+
```
Comment (arnaudsjs) [inline docs/reference/api.rst:9]:
```suggestion
    Only those parts explicitly mentioned here are part of the stable API. Other parts
    of the containing modules provide no such guarantees.

```
The term "internal API" confused me. "internal API" and "stable API" are pointing to the same thing here, so let's choose one term and use it consistently.

**Comment by arnaudsjs** [comment]
The build of the documentation also shows the following warning:
```
inmanta/model.py:docstring of inmanta.model.Attribute:: WARNING: more than one target found for cross-reference 'Location': inmanta.model.Location, inmanta.ast.export.Location
inmanta/model.py:docstring of inmanta.model.Entity:: WARNING: more than one target found for cross-reference 'Location': inmanta.model.Location, inmanta.ast.export.Location
inmanta/model.py:docstring of inmanta.model.Relation:: WARNING: more than one target found for cross-reference 'Location': inmanta.model.Location, inmanta.ast.export.Location
```

File: src/inmanta/ast/type.py (hunk 3 above)
Comment (sanderr) [reply src/inmanta/ast/type.py:89]: This is not a typo. Iff stands for "if and only if"

File: docs/reference/api.rst (hunk 4 above)
Comment (sanderr) [reply docs/reference/api.rst]: I opted to just remove the "internal" keyword. I think the second sentence is required for the third to make sense.

**Pull request merged**